### ISSUE TITLE
Rust Policy

### ISSUE THREAD

**Issue by BryanQuigley** [systemd/systemd#18767]
**Is your feature request related to a problem? Please describe.**
Issue #13135 helped me understand that [zram-generator](https://github.com/systemd/zram-generator) is not in systemd's main repo because it's written in Rust. I'm simply wondering if there was a discussion about what it would take for systemd's main repo to start including Rust.

For reference, both [Linux](https://lore.kernel.org/lkml/CAKwvOdmuYc8rW_H4aQG4DsJzho=F+djd68fp7mzmBp3-wY--Uw@mail.gmail.com/T/#u) and [Mesa](https://lists.freedesktop.org/archives/mesa-dev/2020-October/224639.html) have had such discussions, but I'm unaware of any Rust code that has landed in either. 

**Describe the solution you'd like**
Any clear answer is appreciated, including:
 * Rust can be added under conditions X, Y and Z.
 * Rust can't be considered yet due to A, B.
 * We are a C project

**The systemd version you checked that didn't have the feature you are asking for**
248-rc1

**Comment by keszybz** [comment]
I'd say that we don't want to add rust code, yet. The problem is that systemd is installed on a very wide range of hardware, some of which is not supported by rust. For example the seccomp filters we have are architecture specific, and people have file many issues about various fringe architectures, which means that they are compiling systemd there.

Another issue is code size — right now zram-generator binary package in Fedora is 1MB. It has a whopping total of 1000 lines of source code. It'd be nice to convert some systemd utilities to rust, but it'd mean an explosion in package size. And since our binaries are installed on tiny devices, in initramfs, in containers, etc, that'd be very painful.

But as the situation both with arch coverage and shared-library reuse improves, this can be revisited.

**Comment by vcaputo** [comment]
I wonder if a simple position of "we'll support Rust code in systemd once Linus' linux tree requires Rust to compile" makes sense, since systemd is so linux-coupled.

**Comment by keszybz** [comment]
> I wonder if a simple position of "we'll support Rust code in systemd once Linus' linux tree requires Rust to compile" makes sense, since systemd is so linux-coupled.

I'd prefer to stay independent, though the state of dependencies is certainly something to take into account. I guess that when a major use case comes along, we will have to make a decision.

**Comment by poettering** [comment]
I think before we can adopt Rust there needs to be some reasonable solution for building truly hybrid codebases. Last time I looked Cargo and Meson don't get along too well. systemd is after all a big set of indvidual tools, and if we move to Rust we'd have to do this bit by bit, and thus for a longer time carry both C and Rust code in our codebase, and it's probably going to me mixed not only at the "leaves" of the source dependency trees, but also in the "trunks" of it, i.e the common code in src/shared/ or src/basic. But mixing the two languages in one codebase at multiple places needs a build system that actually can build that reasonably. Last time I looked existing hybrid projects usually have very clear separation that the frontend is one language and backend the other, or that library is one language and the consumer in the other. If you have such simple separation glueing things together hafl-heartedly might work, but if we have a big codebase, consisting of many components then it's going to require some way to build it that isn't just hotglueing stuff that doesn't really fit together.

I'd much prefer if others would build the tool chain infrastructure for this for us, before we start doing Rust, because I don't think we should be pioneering any of that. Having a complex project with a hybrid C+Rust codebase can't be that exotic, and I'd rather wait until someone else fixed the harder problems this require for us, before we do it ourselves. After all we want to focus on different things in systemd development than to make a build system work. once we adopt Rust we want to adopt it as a tool like any other, and not the object of our work itself.

I also kinda hope that other projects will already adress the politics of this, i.e. push the Debians and Gentoos of this world to figure out how they can combine their love of obsolete archs and their interest in a more secure future in some way, or if they have to drop one for the other.

**Comment by poettering** [comment]
(btw, as I understand the Rust efforts in kernel land: they focus on drivers only for rust, i.e. core kernel code is not going to be rust code, afaiu. Limiting things like that means only doing rust in the "leaves" of the dep tree, i.e. building gets much simpler. in systemd that doesn't really map so well however, we have no clear concept of "drivers" vs. "core kernel" or anything like that)

**Comment by poettering** [comment]
I think this has been thorougly answered and we can close this, right?

**Comment by BryanQuigley** [comment]
Indeed, Thank you all for the detailed answers!
